### PR TITLE
ENH: better detection of categorical columns in plot and explore

### PR DESCRIPTION
**Pull request by keirayuki**
Addresses the issue of #2416  
This is my first contributing oss activity.  
So if you find something wrong, please point it out.

The main changes are as follows
- Using pandas API instead of numpy API
- Adding conditions in plotting as well as explore  

The reason for the above changes is that using pandas.api.types can be better way to catch more conditions, as stated in #2416.

### CODE REVIEW

**Comment by keirayuki** [comment]
Thanks for the comments!  
I've submit some test cases and additionally the way to deal with string case.  

In the following case, category detection wouldn't work.  

```python
import numpy as np
import pandas as pd
import geopandas as gpd

gdf = gpd.read_file(gpd.datasets.get_path("nybb"))
gdf["bool"] = [True, False, True, True, False]
gdf["bool_extention"] = pd.array([True, False, True, True, False])
gdf["string"] = pd.array(([True, False, True, True, False]), dtype="string")

gdf.plot("bool") # category detection work well (#2416)
gdf.explore("bool") # category detection work well (#2416)

gdf.explore("bool_extention") # category detection wouldn't work
gdf.plot("string") # category detection wouldn't work
gdf.explore("string") # category detection wouldn't work
```

So, I fixed by using pandas API insetad of numpy api in explore.py and plotting.py like the following.  
```python
    elif values.dtype is np.dtype("O") or categories:
    elif (
        pd.api.types.is_object_dtype(values.dtype)
        or pd.api.types.is_bool_dtype(values.dtype)
        or pd.api.types.is_string_dtype(values.dtype)
        or categories
    ):
```
Also, I've written the test cases to check them out.  

About `pd.api.types.is_obeject_dtype`, maybe, I thougt that `pd.api.types.is_obeject_dtype` was better than `np.dtype('O')`. 
(I'm not unconfident with this and I can't come up with the test cases.)

File: geopandas/tests/test_plotting.py [hunk 1]
```diff
@@ -405,6 +405,11 @@ def test_categories(self):
         ):
             self.df.plot(column="cats", categories=["cat1"])
 
+    def test_string_category_plot(self):
+        df = self.df.copy()
+        df["string"] = pd.array(["1", "2", "3", "4", "5"] * 2, dtype="string")
```
[the remark this answers is not in the document]
Comment (keirayuki) [reply geopandas/tests/test_plotting.py]: Thanks for code review!
I removed test_string_category_plot and moved its test case to test_categories in test_protting.
Also, I added test case about boolean array there.
(I'm sorry if my understanding is different from what you think)

```diff
def test_categories(self):
    self.df["cats_object"] = ["cat1", "cat2"] * 5
    self.df["nums"] = [1, 2] * 5
    self.df["singlecat_object"] = ["cat2"] * 10
    self.df["cats"] = pd.Categorical(["cat1", "cat2"] * 5)
    self.df["singlecat"] = pd.Categorical(
         ["cat2"] * 10, categories=["cat1", "cat2"]
    )
    self.df["cats_ordered"] = pd.Categorical(
        ["cat2", "cat1"] * 5, categories=["cat2", "cat1"]
    )
+   self.df["bool"] = [False, True] * 5
+   self.df["bool_extension"] = pd.array([False, True] * 5)
+   self.df["cats_string"] = pd.array(["cat1", "cat2"] * 5, dtype="string")

    ax1 = self.df.plot("cats_object", legend=True)
    ax2 = self.df.plot("cats", legend=True)
    ax3 = self.df.plot("singlecat_object", categories=["cat1", "cat2"], legend=True)
    ax4 = self.df.plot("singlecat", legend=True)
    ax5 = self.df.plot("cats_ordered", legend=True)
    ax6 = self.df.plot("nums", categories=[1, 2], legend=True)
+   ax7 = self.df.plot("bool", legend=True)
+   ax8 = self.df.plot("bool_extension", legend=True)
+   ax9 = self.df.plot("cats_string", legend=True)

    point_colors1 = ax1.collections[0].get_facecolors()
+    for ax in [ax2, ax3, ax4, ax5, ax6, ax7, ax8, ax9]:
        point_colors2 = ax.collections[0].get_facecolors()
        np.testing.assert_array_equal(point_colors1[1], point_colors2[1])

    legend1 = [x.get_markerfacecolor() for x in ax1.get_legend().get_lines()]
+    for ax in [ax2, ax3, ax4, ax5, ax6, ax7, ax8, ax9]:
        legend2 = [x.get_markerfacecolor() for x in ax.get_legend().get_lines()]
        np.testing.assert_array_equal(legend1, legend2)
```

**Review by martinfleis — APPROVED** [review]
Thanks!

**Review by m-richards — APPROVED** [review]
Thanks @keirayuki, looks good to me as well!

**Pull request merged**